[exporter/azuremonitor] Add resource attributes, instrumentation scope and cloud tags to Azure Monitor exporter for logs

bastbu opened the pull request:
**Description:**

Fixes #18525

The Azure Monitor exporter did not export [resource attributes](https://opentelemetry.io/docs/reference/specification/logs/data-model/#field-resource) and [instrumentation scope](https://opentelemetry.io/docs/reference/specification/logs/data-model/#field-instrumentationscope) as `customDimensions` to Application Insights. To achieve more parity with respect to how traces and metrics are exported, this PR also applies cloud tags by invoking `applyCloudTagsToEnvelope` to the log records.

**Link to tracking Issue:** #18525

**Testing:** Manual testing with Application Insights, automated testing through unit tests.

**Documentation:** No documentation added

bastbu committed:
Set customDimensions equivalent to metrics for logs exported to Azure Monitor

bastbu committed:
Add changelog entry

eedorenko reviewed (approved):
LGTM

eedorenko commented:
Fixes #18525
Fixes #16557

bastbu commented on .chloggen/fix_resource-attributes-logs.yaml:
```diff
@@ -0,0 +1,16 @@
+note: "Ensure that resource attributes, instrumentation scope and cloud tags are exported to Azure Monitor for logs"
+
+# One or more tracking issues related to the change
+issues: [18525]
```
Not yet, when I tested this I saw that structured logging is not yet supported (though it addresses [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16557#issuecomment-1340859883)).

For structured logging, I believe I will need to add all (non-resource) attributes as well. I would propose to do this in a separate PR.

bastbu replied:
Created #18917 to address said issue.

codeboten reviewed (commented):
Pinging codeowner @pcwiese

atoulme commented:
Would you please add a test showing how this changes the export?

bastbu committed:
Add unit tests for export structure updates

bastbu commented:
> Would you please add a test showing how this changes the export?

@atoulme Sure, added the unit tests for how the mapping to the Application Insights contract changes with ef3d956ad3c9e2b7f49aaaff1a6d7d91dadbafb8. Updated the PR description accordingly.

bastbu commented:
@dashpole Would it be possible to merge this PR?

evan-bradley commented on exporter/azuremonitorexporter/logexporter_test.go:
```diff
@@ -128,6 +129,50 @@ func TestExporterLogDataCallback(t *testing.T) {
+	envelope := logPacker.LogRecordToEnvelope(logRecord, getResource(), scope)
+
+	envelopeData := envelope.Data.(*contracts.Data).BaseData.(*contracts.MessageData)
+	require.Equal(t, scope.Name(), envelopeData.Properties[aiInstrumentationLibraryNameConvention])
+	require.Equal(t, scope.Version(), envelopeData.Properties[aiInstrumentationLibraryVersionConvention])
```
I think it would make sense to use the same attribute constants that the production code uses here.

bastbu replied:
To make sure I understand the suggestion correctly: the suggestion would be to import the constant `instrumentationLibraryVersion` from `contract_utils.go`, correct?

My thinking behind "duplicating" the constants is that if someone changes the value of `instrumentationLibraryVersion` in the production code, which is a breaking change, there will be a test that breaks.

Hence I would tend to keep it as it is, but no strong opinion here and happy to apply your suggestion if you're not convinced.

evan-bradley replied:
That's a fair point. I don't feel strongly about it either, so I'm good leaving it as-is.

evan-bradley commented on exporter/azuremonitorexporter/logexporter_test.go:
```diff
@@ -128,6 +129,50 @@ func TestExporterLogDataCallback(t *testing.T) {
+
+	resourceAttributes := resource.Attributes().AsRaw()
+	expectedCloudRole := resourceAttributes[conventions.AttributeServiceNamespace].(string) + "." + resourceAttributes[conventions.AttributeServiceName].(string)
+	require.Equal(t, expectedCloudRole, envelope.Tags[aiCloudRoleConvention])
+	expectedCloudRoleInstance := resourceAttributes[conventions.AttributeServiceInstanceID]
+	require.Equal(t, expectedCloudRoleInstance, envelope.Tags[aiCloudRoleInstanceConvention])
```
I think it would make sense to use the same attribute constants that the production code uses here as well.

evan-bradley replied:
Based on the discussion on the instrumentation scope constants, I'm okay leaving this as-is.
